File: csharpcore/TellDontAskKata/TellDontAskKata.Main/UseCase/OrderShipmentUseCase.cs
using TellDontAskKata.Main.Domain;
using TellDontAskKata.Main.Repository;
using TellDontAskKata.Main.Service;

namespace TellDontAskKata.Main.UseCase
{
    public class OrderShipmentUseCase
    {
        private readonly IOrderRepository _orderRepository;
        private readonly IShipmentService _shipmentService;

        public OrderShipmentUseCase(
            IOrderRepository orderRepository,
            IShipmentService shipmentService)
        {
            _orderRepository = orderRepository;
            _shipmentService = shipmentService;
        }

        public void Run(OrderShipmentRequest request)
        {
            var order = _orderRepository.GetById(request.OrderId);

            if (order.Status == OrderStatus.Created || order.Status == OrderStatus.Rejected)
            {
                throw new OrderCannotBeShippedException();
            }

            if (order.Status == OrderStatus.Shipped)
            {
                throw new OrderCannotBeShippedTwiceException();
            }

            _shipmentService.Ship(order);

            order.Status = OrderStatus.Shipped;
            _orderRepository.Save(order);
        }
    }
}

File: scala/src/main/scala/ordershipping/usecase/OrderShipmentUseCase.scala
package ordershipping.usecase

import ordershipping.domain.OrderStatus
import ordershipping.repository.OrderRepository
import ordershipping.service.ShipmentService

class OrderShipmentUseCase(
    val orderRepository: OrderRepository,
    val shipmentService: ShipmentService
) {
  def run(request: OrderShipmentRequest): Unit = {
    orderRepository
      .getById(request.orderId)
      .foreach(order => {
        if (
          order.status == OrderStatus.Created ||
          order.status == OrderStatus.Rejected
        ) throw new OrderCannotBeShippedException

        if (order.status == OrderStatus.Shipped)
          throw new OrderCannotBeShippedTwiceException

        shipmentService.ship(order)
        order.status = OrderStatus.Shipped
        orderRepository.save(order)
      })
  }
}
